tmexp/create_bow.py

```diff
@@ -0,0 +1,218 @@
+from collections import Counter
+import copy
+import logging
+import os
+import pickle
+from typing import Any, Dict, List, Optional
+
+import tqdm
+
+from .utils import (
+    check_exists,
+    check_remove_file,
+    create_directory,
+    create_language_list,
+)
+
+DIFF_MODEL = "diff"
+HALL_MODEL = "hall"
+SEP = ":"
+
+
+def check_fraction(fraction: float, arg_name: str) -> None:
+    if not (0 <= fraction <= 1):
+        raise RuntimeError(
+            "Argument '%s' must be in the range [0, 1], aborting." % arg_name
+        )
+
+
+def create_bow(
+    input_path: str,
+    output_dir: str,
+    dataset_name: str,
+    langs: Optional[List[str]],
+    exclude_langs: Optional[List[str]],
+    features: List[str],
+    force: bool,
+    topic_model: str,
+    min_word_frac: float,
+    max_word_frac: float,
+    log_level: str,
+) -> None:
+    logger = logging.getLogger("create_bow")
+    logger.addHandler(logging.StreamHandler())
+    logger.setLevel(log_level)
+
+    check_exists(input_path)
+    create_directory(output_dir, logger)
+    if dataset_name == "":
+        dataset_name = topic_model
+    words_output_path = os.path.join(output_dir, "vocab." + dataset_name + ".txt")
+    check_remove_file(words_output_path, logger, force)
+    docword_output_path = os.path.join(output_dir, "docword." + dataset_name + ".txt")
+    check_remove_file(docword_output_path, logger, force)
+    docs_output_path = os.path.join(output_dir, "docs." + dataset_name + ".txt")
+    check_remove_file(docs_output_path, logger, force)
+    check_fraction(min_word_frac, "min-word-frac")
+    check_fraction(max_word_frac, "max-word-frac")
+
+    logger.info("Reading pickled dict from '%s' ..." % input_path)
+    with open(input_path, "rb") as _in:
+        input_dict = pickle.load(_in)
+
+    logger.info("Computing bag of words ...")
+    langs = create_language_list(langs, exclude_langs)
+    all_refs = sorted([ref for ref in input_dict["files_info"]])
+    doc_freq: Counter = Counter()
+    bow: Dict[str, Dict[int, Dict[str, Any]]] = {}
+    num_docwords = 0
+    num_blobs = 0
+    for file_path, blob_dict in tqdm.tqdm(input_dict["files_content"].items()):
+        refs = sorted(
+            [
+                ref
+                for ref, file_dict in input_dict["files_info"].items()
+                if file_path in file_dict
+            ]
+        )
+        lang = input_dict["files_info"][refs[0]][file_path]["language"]
+        if lang not in langs:
+            continue
+        word_counts = {}
+        for blob_hash, feature_dict in blob_dict.items():
+            word_dict: Counter = Counter()
+            for feature in features:
+                if feature not in feature_dict:
+                    continue
+                word_dict.update(feature_dict[feature])
+            if word_dict:
+                num_blobs += 1
+                doc_freq.update(word_dict.keys())
```

Comment:
Sorry I did not review the logic in-depth because it's quite intricate as always with feature extraction, but doesn't the `.keys()` wrongly set to 1 the counts of each word? Or is it wanted?

Reply:
No problem, it does get a bit complex in that area. It is wanted, because if I don't use  `.keys()`, since `word_dict` is also a `Counter` then for each word we will increment by the number of occurrences in the blob, however `doc_freq` tracks the number of docs a word is in, not the total amount of times it appears in the corpus.

Reply:
thanks for clearing it up :pray: